### ISSUE TITLE
Question: Why does `UnitCell::cal_ux()` rely on the unchanged `atoms[it].m_loc_[ia]`?

### ISSUE THREAD

**Issue by WHUweiqingzhou** [deepmodeling/abacus-develop#4040]
### Details

During SCF,  `atoms[it].m_loc_[ia]` is not updated, and `cal_ux` run at each iteration. Is it correct?

### Have you read FAQ on the online manual http://abacus.deepmodeling.com/en/latest/community/faq.html

- [X] Yes, I have read the FAQ part on online manual.

### Task list for Issue attackers (only for developers)

- [ ] Understand the problem or question described by the user.
- [ ] Check if the issue is a known problem or has been addressed in the documentation.
- [ ] Test the issue or problem on a similar system or environment, if possible.
- [ ] Identify the root cause or provide clarification on the user's question.
- [ ] Provide a step-by-step guide, including any necessary resources, to resolve the issue or answer the question.
- [ ] If the issue is related to documentation, update the documentation to prevent future confusion (optional).
- [ ] If the issue is related to code, consider implementing a fix or improvement (optional).
- [ ] Review and incorporate any relevant feedback from users or developers.
- [ ] Ensure the user's issue is resolved or their question is answered and close the ticket.

**Comment by mohanchen** [comment]
Good question.

**Comment by dyzheng** [comment]
This function only used for the special case: `antiferromagnetism with NSPIN=4`

In most of case, the local magnetic moment on grid point can be defined as `mag(r) = (rho_x, rho_y, rho_z)`, and when rotate the quantization axis toward `mag'(r) = (0, 0, rho'_z)`,  the `rho'_z` can be `sqrt(|mag(r)|^2)`, which is always positive.

But error occurred with `antiferromagnetism` case that some local magnetic moment on different grid point would have same norm with opposite direction, which usually will not arise naturally without initial antiferromagnetism atomic settings. In this case, `cal_ux()` will define a positive direction, and the `antiferromagnetism` will be calculated correctly.

In conclusion, this method for calculating magnetization density with noncollinear spin can work, but it dependent on the user's initial magnetic moment Settings, very unreliable!!! 

I plan to refactor this method in future.

**Comment by WHUweiqingzhou** [comment]
@dyzheng 
Thanks! I get it.
